### PR TITLE
Curve curve intersection, take two

### PR DESCRIPTION
**Pull request by simoncozens**
This borrow's lyon's curve-curve intersection algorithm. Some notes:

* I've tried to separate out the sections of the algorithm which are cubic-specific into a separate trait, leaving the generic parts in `curve_intersections`. 
* In theory, this means that the trait can also be implemented on a `QuadBez`, and we would get cubic/quad and quad/quad intersections for free. In practice, I don't understand the algorithm well enough to implement the convex hull thingy on quads.
* It also means that the new `ParamCurveBezierClipping` trait is a bit of a grab-bag of whatever methods were needed to get the algorithm working. Some of these methods (particularly `solve_t_for_x` and `solve_t_for_x`) may want to go into other traits - but which traits?

Again, for discussion.

### CODE REVIEW

**Comment by tylers-epilog** [comment]
One thing to note that might be useful here is that quad beziers can easily be converted into cubic beziers. I haven't started looking through the code, so I'm not sure if a function for this exists already, but if not, take a look at this:
https://stackoverflow.com/a/3162732

**Review by tylers-epilog — COMMENTED** [review]
Overall, it looks good. I'm not a contributor yet, so I'm not sure that I should change the status of this PR (approval or change-request), but I would recommend considering some of my ideas before merging this PR since I've gone through this before and the edge cases can be killer.

Also, let me know if you need/want me to implement the changes that I've requested, or if you have questions about them.

**Comment by simoncozens** [comment]
Thanks for this! Please note that the whole implementation, comments-and-all, is lifted from Lyon: https://github.com/nical/lyon/blob/master/crates/geom/src/cubic_bezier_intersections.rs

So if there are improvements, maybe they should head upstream as well?

**Comment by tylers-epilog** [comment]
That's a good question. It looks like lyon is still being actively developed, so I suppose they should head upstream as well. We should test the improvements thoroughly first, but those devs may also have some insight into the changes we're making, which means we'll get even more eyes on it.

File: src/curve_intersections.rs [hunk 1]
```diff
@@ -0,0 +1,604 @@
...
+        let curve_t = domain_value_at_t(curve_domain, t);
+        add_intersection(pt_t, pt_curve, curve_t, curve, flip, intersections);
+    }
+}
+
+fn point_curve_intersections<T: ParamCurveBezierClipping>(
+    pt: Point,
+    curve: &T,
+    epsilon: f64,
+) -> ArrayVec<[f64; 9]> {
```
[the remark this answers is not in the document]
Comment (tylers-epilog) [reply src/curve_intersections.rs:300]: If you don't mind, I think I'll have a crack at implementing (and testing) this. I'll try to make the function as general as possible so that it can be useful to other parts of the code as well. When I'm done I'll pull request it into your branch so that you can review it.

**Comment by tylers-epilog** [comment]
Hey @simoncozens. I just wanted to check in and see if you saw the pull requests hat I've made into your repo. I thought it might be best to do it that way so that this pull request remains valid but is updated with the changes that I've made.

**Comment by simoncozens** [comment]
I had not! I shall take a look.

**Comment by simoncozens** [comment]
You know what, I think I should just be totally honest here: I really want this functionality to exist; but I also really don't want to spend much more of my limited programming time on nasty algorithmic geometry edge-cases. I thought I could get away with grabbing the code from Lyon and making it work, but if people are going to want me to actually *understand* it and *improve* it, I don't think I'm ever going to find the time or headspace to do so.

So if anyone wants to take over this PR, please do so.

**Comment by raphlinus** [comment]
I'm in a similar boat, I want this functionality in kurbo, but also don't have the bandwidth to wrap my head around it. So here's what I think we should do:

Merge this code basically as-is, with a disclaimer added to the doc comment that it hasn't been validated thoroughly for edge cases and may not be numerically robust. Then I'll file an issue for the robustification. I have some thoughts on this topic; it's probably a valid research topic in 2D computational geometry, and to my knowledge a numerically robust cubic-cubic intersection problem hasn't been published. Ideally that would attract somebody interested in the topic, or, at some point, I'll probably have a whack at it myself.

Sound good?

**Comment by tylers-epilog** [comment]
I'm working on this project because I need this algorithm for my job. I've done this before in c++, so I have quite a bit of experience with the edge cases (and I'm not quite sure that you can ever make this algorithm 100% robust). As I've mentioned to Simon, my main interest is in creating boolean path operations (e.g. union of two closed paths, etc) and Kurbo is a good base for that project. If this doesn't interest others, then I'll just create my own fork/project and work from there. However, I think the others in the community would probably benefit from this project.

@simoncozens and @raphlinus If what I've said above sounds good to you, I'm happy to take over this pull request. I've posted 2 pull-requests into this branch just so that you can see what I've been doing, but if I take over this pull request, I'll just merge them in.

Let me know what you think!

**Comment by tylers-epilog** [comment]
I guess it depends on what you mean by numerically robust. There's always going to be some amount of floating-point error. However, I do think that the improvements I've made make the accuracy pretty good.

I think that preventing regression from Lyon _might_ be an issue. Mainly due to the fact that they have a bunch of random epsilon values and they don't capture intersections that happen at the endpoints. In the work that I've done, I've added an input flag that allows developers to choose whether they want to capture endpoint intersections or not, but the definition of an "endpoint" is of course a little fuzzy. I'm trying to make my code as useful to as many people as possible, but I don't know what constraints/edge cases were used in making some of their design choices. As such, I'm not sure that I can make it work for my project AND make it 100% compatible with Lyon unless I have them test my changes. Maybe my code would work for them, maybe it wouldn't. What do you think is the right course of action in this case?

I'm totally fine with merging this in and then posting a new pull request with my changes.

**Comment by raphlinus** [comment]
The first step in numerical robustness is coming up with a good definition of what that means. That should probably be done in the context of the task, in this case boolean path operations. A reasonable first cut is that the curves must be close (within some epsilon) at the intersection points, and the relative orientations must be consistent (this would need to be precisely defined, I'm just sharing my intuition). It's possible that for "relative orientation" to be well defined, the API should be changed so that it reports at least whether the order of intersection is odd or even, ie whether there is a sign change or not (y = x^2 intersect y = 0 is an example of a second-order intersection, and I believe we can get up to fourth-order by intersecting two cubics). Another approach is to detect the numerically unstable cases and upgrade to higher precision arithmetic; this is what Shewchuk does in his [Triangle](https://www.cs.cmu.edu/~quake/triangle.html) project. I'd prefer not to do that and think it's not necessary, but that approach shows it is possible.

But let's take things in order. Let's land this PR with minimal changes and a disclaimer (it sounds like Tyler's work also makes minor changes to the API), then I'll set aside some time to review Tyler's PR. I also have some ideas on how to generate adversarial examples.

**Comment by tylers-epilog** [comment]
That sounds good. I guess that just means that @simoncozens needs to make a commit adding a disclaimer that this code hasn't been tested against every edge case?

Once my PR is up, we can discuss the numerical robustness of my code and how to go about any improvements that you think can be made since you seem to have a pretty good grasp on the subject. We can also then discuss some of the adversarial examples that you have in mind and you can compare them to the test cases that I've added.

**Comment by simoncozens** [comment]
That all works for me. I'll add a doc disclaimer in a bit.

**Comment by raphlinus** [comment]
I wrote a lot more thoughts on this on the Zulip: https://xi.zulipchat.com/#narrow/stream/260979-kurbo/topic/B.C3.A9zier-B.C3.A9zier.20intersection

**Comment by tylers-epilog** [comment]
Aside from some of it being _way_ over my head :), I like what you've written down and I think you make some good points. Getting a numerically robust algorithm now will save us a lot of time later and will prevent having to code for edge cases. In particular, I'd like to see your work on calculating the Hausdorff distance. 

In my experience of the path operations, two curves lying very close to each other is one of the biggest issues. In my code that I'll post, I've added a check to see if two curves "lie on top" of each other, but your Hausdorff method is probably a better check.
